### PR TITLE
Fixed #20488 -- Added copy and move methods to the storage API

### PR DESCRIPTION
**Pull request by kux**
* implemented these methods in the default FileSystemStorage
* fix issue with storage tests: each test from FileStorageTests
   was run twice (when inheriting from a TestCase class all test_
   methods are ran in the subclass as well)

Previous incarnations of the same PR:

https://github.com/django/django/pull/1206
https://github.com/django/django/pull/2664

### CODE REVIEW

**Comment by collinanderson** [comment]
buildbot, test this please

File: tests/file_storage/tests.py [hunk 1]
```diff
@@ -405,6 +408,48 @@ def test_delete_no_name(self):
...
+        with self.assertRaises(SuspiciousOperation):
+            self.storage.move(src_name, '/tmp/passwd')
+
+    def test_move_file_no_directory_traversal(self):
+        f = ContentFile('content content')
+        src_name = self.storage.save('initial.file', f)
+        with self.assertRaises(SuspiciousOperation):
+            self.storage.move(src_name, '../../../../../../../../tmp/passwd')
+
+
```
Comment (collinanderson) [inline tests/file_storage/tests.py]: Remove one of these blank lines (should only be two total). (`tests/file_storage/tests.py:454:1: E303 too many blank lines (3)`)

File: django/core/files/copy.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import os
+from shutil import copystat
+
+from django.core.files import locks
+
+
+__all__ = ['file_copy_safe']
+
+
+def file_copy_safe(old_file_name, new_file_name, chunk_size=1024*64, allow_overwrite=False):
```
Comment (collinanderson) [inline django/core/files/copy.py]: `E226 missing whitespace around arithmetic operator`

File: django/core/files/copy.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    Copy a file from one location to another in the safest way possible.
+
+    If the destination file exists and ``allow_overwrite`` is ``False``, this
+    function will throw an ``IOError``.
+    """
+    # first open the old file, so that it won't go away
+    with open(old_file_name, 'rb') as old_file:
+        # now open the new file, not forgetting allow_overwrite
+        fd = os.open(new_file_name, os.O_WRONLY | os.O_CREAT | getattr(os, 'O_BINARY', 0) |
+                                    (not allow_overwrite and os.O_EXCL or 0))
```
Comment (collinanderson) [inline django/core/files/copy.py]: `django/core/files/copy.py:28:37: E127 continuation line over-indented for visual indent`
pep8 wants you to indent it like this:

    fd = os.open(new_file_name, os.O_WRONLY | os.O_CREAT | getattr(os, 'O_BINARY', 0) |
                 (not allow_overwrite and os.O_EXCL or 0))

**Comment by kux** [comment]
Closed in favor of rebased https://github.com/django/django/pull/3915

**Pull request closed without merging**